### PR TITLE
I13 ekf slam

### PR DESCRIPTION
**Pull request by nanwei1**
Files are updated in i13/code v1.0/ME597-6-EKFSLAM

Changes made:
- Functionalized measurement and motion models
- Linearized models were also functionalized
- The main difference from the previous version (ekf_slam) is that the map size is updated as new features are detected:
  - Initially the states and covariance only contain information about the vehicle (x,y,theta in this case)
  - As new features are observed, the estimate of this feature's state is augmented right below the vehicle states. The feature list, measurements, etc are rearranged accordingly
  - The covariance for the newly observed feature is set to a predefined value (S0mm=eye(2)), and the off-diagonal terms are set to 0
  - A line of text is printed to the Command Window, indicating how many features have been detected
-  Added examples of vehicle running with different trajectories in different maps (ekf_slam_lineMap and ekf_slam_circularMap)
- The EKF function will be incorporated in the next round
- Note that just like the original code, this simulation fails from time to time as the linearization may not be accurate or the covariance matrix becomes not positive-definite. This code warns the user when the update in vehicle position is too large compared to input velocity but cannot deal with these failures. Any comments on how to fix this will be appreciated.
- Added example where features are placed further way from vehicle trajectory (ekf_slam_largerCircularMap) to show that linearization works better in this case (as illustrated by Mapping II slide 28). If you try these simulation enough times, you will see that this one fails less frequently than ekf_slam_circularMap. Below is a comparison of the estimates:
  radius 6
  (Features in a r=6m circle)

radius 12
(Features in a r=12m circle)

### CODE REVIEW

**Comment by cbchau** [comment]
Suggest putting the plotting code in separate functions which can be used for multiple scripts.
plot code

**Comment by nanwei1** [comment]
- Combined all 4 examples into 1 file. The user will need to pick an example at the beginning of the code.
- An ekfSLAMplot function is used to replace the original ~40 lines for plotting
- In case if S (the covariance matrix) gets not positive definite, manually add a diagonal matrix to make it positive definite to keep the code running. Warn the use when this happens
- I considered using the ekf function, but the issue was that this ekf_slam code updates the size of its map and covariance matrix as new features are discovered. Initialization of newly discovered features is performed after the prediction step using the predicted robot states, and the size of mu and S is updated before the measurement update. Thus, I did not find a way to directly use the ekf function without much modification
- Range/bearing only measurement model was tested. I did not upload the code because it basically did not work. The problem began with initialization of features; with only 1 measurement, we needed to estimate the (x,y) position of a feature, which results in poor initialization that messed up the whole simulation

File: matlab_simulation/code v1.0/ME597-6-EKFSLAM/ekf_slam.m [hunk 1]
```diff
@@ -113,102 +116,94 @@
...
+            Ht = range_bearing_meas_linearized_model(mu,j);
+            I = y(2*(j-1)+1:2*j,t)- range_bearing_meas_model(mu(1:3),mu((3+2*(j-1)+1):3+2*j));
  
             % Measurement update
             K = S*Ht'*inv(Ht*S*Ht'+Qi);
             mu = mu + K*I;
-            S = (eye(n+2*M)-K*Ht)*S;
+            S = (eye(length(S))-K*Ht)*S;
+            
+            % In cases if S bemoes not positive definite, manually make it
```
Comment (chutsu) [inline matlab_simulation/code v1.0/ME597-6-EKFSLAM/ekf_slam.m:187]: minor edit, suggest 'is not positive definite' over 'becomes not positive definite'

File: matlab_simulation/code v1.0/ME597-6-EKFSLAM/range_bearing_meas_linearized_model.m [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+% linearized measurement model of range and bearing
+% can be found P.22 of Mapping II slides
+            dx = mu(3+2*(i-1)+1)-mu(1);
+            dy = mu(3+2*i)-mu(2);
+            rp = sqrt((dx)^2+(dy)^2);
+            
+            N=length(mu);
+            Fi = zeros(5,N);
+            Fi(1:3,1:3) = eye(3);
+            Fi(4:5,3+2*(i-1)+1:3+2*i) = eye(2);
```
Comment (chutsu) [inline matlab_simulation/code v1.0/ME597-6-EKFSLAM/range_bearing_meas_linearized_model.m:11]: some explaination of how this `Fi` is used would be nice.

File: matlab_simulation/code v1.0/ME597-6-EKFSLAM/range_bearing_meas_linearized_model.m [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+                dx/rp ...
+                dy/rp;
+                dy/rp^2 ...
+                -dx/rp^2 ...
+                -1 ...
+                -dy/rp^2 ...
+                dx/rp^2]*Fi;
+            
+             h_mu_bar = [rp;
+                (atan2(dy,dx) - mu(3))];
```
Comment (chutsu) [inline matlab_simulation/code v1.0/ME597-6-EKFSLAM/range_bearing_meas_linearized_model.m:24]: `h_mu_bar` is not used anywhere.

File: matlab_simulation/code v1.0/ME597-6-EKFSLAM/two_wheel_motion_linearized_model.m [hunk 4]
```diff
@@ -0,0 +1,10 @@
+function Gt = two_wheel_motion_linearized_model(x,u,dt)
+% linearized motion model of a 2-wheel robot
+% can be found P.20 of Mapping II slides
```
Comment (chutsu) [inline matlab_simulation/code v1.0/ME597-6-EKFSLAM/two_wheel_motion_linearized_model.m:3]: Avoid referencing slides, they can change

File: matlab_simulation/code v1.0/ME597-6-EKFSLAM/range_bearing_meas_linearized_model.m (hunk 2 above)
Comment (chutsu) [inline matlab_simulation/code v1.0/ME597-6-EKFSLAM/range_bearing_meas_linearized_model.m:11]: like so:

```
% feature selector vector
Fi = zeros(5, N);   % N is number of elements (num of vehicle states + 2 * num features)
Fi(1:n, 1:n) = eye(n);  % robot x, y, z
Fi(4:5, 3+2*(i-1)+1:3+2*i) = eye(2);  % measurement y1, y2
```

Comment (stevenwaslander) [inline matlab_simulation/code v1.0/ME597-6-EKFSLAM/range_bearing_meas_linearized_model.m:11]:
```
Fi(1:n, 1:n) = eye(n);  % robot x, y, theta
Fi(4:5, 3+2*(i-1)+1:3+2*i) = eye(2);  % feature x, y
```

**Pull request merged**